### PR TITLE
Filippo assessment week1

### PR DESCRIPTION
**Pull request by filippo20u**
last few morning changes before class, challenges not finished

### CODE REVIEW

File: code-challenges.js [hunk 1]
```diff
@@ -14,6 +16,26 @@ var temp2 = 350
...
+
+const tempera = (far) => {
+  if (far === 212){
+    console.log("212 is at the boling point")
+}   else if (far > 212) {
+    console.log("350 is above boiling point")
+}   else if (far < 212) {
+    console.log("35 is below boiling point")
+}
+}
```
Comment (andeeisaacs) [inline code-challenges.js:31]: This is pretty good! Your function is working well, but the problem is your returns. Right now, they are hard coded. What if we were to pass in a temp like.. "500" into your function? It would return "350 is above boiling point." 

Instead, what we could do is utilize string interpolation! 

Additionally, you are using console.log as your return, when you need to be using RETURN. Right now, your function would never produce actual value from your function, it would just log information to the developer's console, but not to anywhere else in the program.

File: code-challenges.js [hunk 2]
```diff
@@ -14,6 +16,26 @@ var temp2 = 350
...
+//just a reminder of the three variables
+console.log(`temp1 = ${temp1}, temp2 = ${temp2}, temp3 = ${temp3}`);
+
+//log the function for the 3 variables
+tempera(temp1)
+tempera(temp2)
+tempera(temp3)
```
Comment (andeeisaacs) [inline code-challenges.js:38]: Here is where we would actually utilize the console log instead. 
`console.log(tempera(temp1))`

File: code-challenges.js [hunk 3]
```diff
@@ -22,6 +44,12 @@ var temp3 = 212
 
 var myNumbers1 = [3, 7, 0, 6, -9]
 
+//I will write a basic loop that will do a basic math function (*5) to each element of the array
+
+for(let i=0; i<myNumbers1.length; i++){
+  console.log(myNumbers1[i] * 5)
+}
```
Comment (andeeisaacs) [inline code-challenges.js:51]: Pretty good! But unfortunately this needs to be wrapped in a FUNCTION as the challenge calls for. Right now, the for loop is also hard-coded, meaning it will only ever work for that variable, myNumbers1.

File: code-challenges.js [hunk 4]
```diff
@@ -31,6 +59,12 @@ var myNumbers1 = [3, 7, 0, 6, -9]
 
 var myNumbers2 = [8, -7, 0, 6, 2]
 
+//I can use the .map syntax to simplify the code, just gett the array values and puty the math operator
+const multi = myNumbers2.map(value => value*5)
```
Comment (andeeisaacs) [inline code-challenges.js:63]: This also needs to be wrapped in a function.

File: code-challenges.js [hunk 5]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
 var stringWithVowels2 = "ILoveJavaScript"
 
 
+//create a new array from called split1 and split2
+//from strings stringWithVowels1 and stringWithVowels2
+
+var split1 = stringWithVowels1.split("")
+var split2 = stringWithVowels2.split("")
```
Comment (andeeisaacs) [inline code-challenges.js:83]: These should exist within the function. As of right now, the function only serves half of its actual purpose, when the function should encompass everything.

File: code-challenges.js [hunk 6]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+      newArray.push(array[i])
+    }
+  }
+  return newArray
+}
+
+// run my function "onlyVow" for the two variables and add a .join accessors to convert it to a strings
+// i can also define a variable and write the .join in a new line
+var combined1 = onlyVow(split1).join("")
+var combined2 = onlyVow(split2).join("")
```
Comment (andeeisaacs) [inline code-challenges.js:107]: These should also be living within the function.

File: code-challenges.js [hunk 7]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+const onlyVow = (array) => {
+  let newArray = []
+  for(let i=0; i<array.length; i++){
+    if(
+       array [i] !== "a"
+    && array [i] !== "e"
+    && array [i] !== "i"
+    && array [i] !== "o"
+    && array [i] !== "u"
+   ){
```
Comment (andeeisaacs) [inline code-challenges.js:97]: Great! But, I would encourage you to format these a bit differently. This is pretty unusual formatting and can become a bit confusing as to what is going on at first glance. 

It should look something more like:

```javascript
if(array[i] !=== "a" && array[i] !=== "e"...
```

File: code-challenges.js [hunk 8]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+}
+
+// run my function "onlyVow" for the two variables and add a .join accessors to convert it to a strings
+// i can also define a variable and write the .join in a new line
+var combined1 = onlyVow(split1).join("")
+var combined2 = onlyVow(split2).join("")
+
+//log the two variables to see the result in the terminal
+console.log(combined1);
+console.log(combined2);
```
Comment (andeeisaacs) [inline code-challenges.js:111]: These console.log's should be calling the function, and passing in the test variables. 

For example:
```javascript
console.log(onlyVow(stringWithVowels1))
console.log(onlyVow(stringWithVowels2))
```

File: code-challenges.js [hunk 9]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
 // --------------------5) Copy/paste your code from #4. Refactor your code to include non-string edge cases. Inform your user that the variable passed into the vowel removal function is not a string.
 // Use the test variables provided below. Expected output: "true is not a string" "42 is not a string"
 
+console.log("Question 5: Unfinished");
+
 var notAString1 = true
 var notAString2 = 42
 
+var split3 = [notAString1]
+var split4 = [notAString2]
```
Comment (andeeisaacs) [inline code-challenges.js:125]: This isnt really doing anything for you.  What these lines are currently saying is:
split3 is an array with true as its only value of the array.

Theres no need to assign the notAString variables to a different datatype.

File: code-challenges.js [hunk 10]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
 
+//define a test variable with multiple type of data 
+var test = [true, false, 42, "HeyThereLearnStudent", 'H', 'e', 'y', 'T', 'h','e', 'r', 'e', 'L', 'e','a', 'r', 'n', 'S', 't','u', 'd', 'e', 'n', 't', ]
+
+//I removed all the conditions for the vowels because I wansn't able to make it work, I tried many possible solutions with no success.
+// I haven't been able to add an else if or else or get the newArray to work as in question 4.
```
Comment (andeeisaacs) [inline code-challenges.js:131]: I see what you are trying to accomplish here, but I think your original code is right. I think you should look into 

[typeof](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/typeof)
[.filter](https://www.w3schools.com/jsref/jsref_filter.asp)

You're really close to finishing this! I know you saw a good example during the morning review.

File: code-challenges.js [hunk 11]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+const hmw5 = (array) => {
```
Comment (andeeisaacs) [inline code-challenges.js:132]: This function should be taking in. as a parameter, a string, not an array.

File: code-challenges.js [hunk 12]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+  [
+  { name: "Itchy", animal: "mouse" },
+  { name: "Stimpy", animal: "cat" },
+  { name: "Daffy", animal: "duck" },
+  { name: "Scratchy", animal: "cat" },
+  { name: "Ren", animal: "dog" },
+  { name: "Felix", animal: "cat" }
+  ]
 
+  var cat = toonimals.filter(value => value.animal === "cat")
```
Comment (andeeisaacs) [inline code-challenges.js:164]: Great! But.. this needs to be wrapped in a function. Right now this is hardcoded to toonimals, we could not pass any other array in.

File: code-challenges.js [hunk 13]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+  { name: "Stimpy", animal: "cat" },
+  { name: "Daffy", animal: "duck" },
+  { name: "Scratchy", animal: "cat" },
+  { name: "Ren", animal: "dog" },
+  { name: "Felix", animal: "cat" }
+  ]
+
+  //trying to run an array destructring to write a function to get only
+  console.log("destru:");
+  var destru =  { name, animal } = toonsDes
```
Comment (andeeisaacs) [inline code-challenges.js:185]: I see you're trying to destructure here, but `name` and `animal` mean nothing to our code right now. name and animal only exist to the local scope of the objects inside of the original array. Not 100% sure how this code is supposed to work for you.

File: code-challenges.js [hunk 14]
```diff
@@ -42,27 +76,142 @@ var stringWithVowels1 = "HeyThereLearnStudent"
...
+
+
+//  2nd attempt, it doesn't filter out the cats.
+// there's an undefined that I'm a little confused where is coming from, i tried to comment out some lines with no success.
+
+const trio = (array) => {
+  array.map(value => {
+    console.log(value.name);
+  })
+}
```
Comment (andeeisaacs) [inline code-challenges.js:203]: Lets remember what every function requires in order for it to work: a return! 

Your function, does not have a return at this time. Your .map ALSO does not have a return yet either. 

It should look, though this code does not output what the challenge is asking for, like this:
```javascript
 const trio = (array) => {
    return array.map(value => {
      return value.name
    })
  }

OR

 const trio = (array) => {
   let nameArray =  array.map(value => {
      return value.name
    })
 return nameArray
  }
```

So this solves your undefined issue, but it still needs to do something more to get the output you need.

File: interview-questions.md [hunk 15]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
 1. What is a function? Why would you use one?
 
   Your answer:
+ A function is a piece of code that acts on other parts of code to do some action on my code like mathematical operation, built in methods etc.
+ I would use one because it's what actually makes my code work and do something rather than just store data.
```
Comment (andeeisaacs) [inline interview-questions.md:12]: Kind of- A function is a piece of code that is able to manipulate data and be reusable. I know you have it better defined in your researched answer :)

File: interview-questions.md [hunk 16]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
+  Functions are useful because I can define the code once, and use it many times with different arguments, to produce different results.
 
 2. What is the difference between .map() and .filter()?
 
   Your answer:
 
-  Researched answer:
+  They are both built-in methods for javascript, .map scans through an array and gives back an array with the same amount of data.
+  .filter scans through an array and not ncesessarly gives back the same amount of data, filtering out what doesn't follow the rules of the conditions of the .filter
```
Comment (andeeisaacs) [inline interview-questions.md:26]: Pretty much! Good work!

File: interview-questions.md [hunk 17]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+  Researched answer:
 
+  .map() and .filter() are two higher-order functions and both have three built-in arguments - the value, index, and array.
+  they both scan through an array, .map returns a new array of the same length, .filter gives a subset of the given array containing only the values that satisfy your stated condition.
 
 3. What is the difference between console.log() and return?
 
   Your answer:
+  console.log is a way in javascript to see what is happening to my code in the terminal. I can see how my data are changing and have a reference of what I'm doing or simply get a message when I run my code.
+  Return is associated to functions and it's part of its syntax, I can have only one return in a function but there's no limimtations for control.log
```
Comment (andeeisaacs) [inline interview-questions.md:37]: Just about! Think of Console.log as your tool to find bugs or examine what something does in your code. It should not be used as an output of a function as it does not 'return' anything. 

Return is what gives our functions and methods an actual output.

File: interview-questions.md [hunk 18]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
...
 
+  console.log() can be called as many times as we want and we can even  inside functions to see what is my output in that specific point
+  Return is part of a function syntax and it's needed to call the function, following the instructions to perform a task but itself it will not give an output in the terminal, to see the output I need to console log the function.
 
 
 4. In regards to functions, what is an argument?
 
   Your answer:
 
+ argument is the data the function is acting on.
```
Comment (andeeisaacs) [inline interview-questions.md:49]: Totally. Its what we pass into the function for the function to manipulate in someway.

File: interview-questions.md [hunk 19]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+  Notice that in the function we created a placeholder called name. This allows us to pass any "data" we want through the function during the function call. the "data" can be a placeholder or can contains other information like variables.  
+ 
+
 
 
 5. Give a brief description of proper pair programming techniques. What are the roles of each person?
 
   Your answer:
 
+ In pair programing you have two people working on the same code, the two share the rationale and one writes the code and the second person double checks if it's well written, there are no typos and gives feedback and suggestions. It's a way to mutually help in case of mistakes and doubts.
```
Comment (andeeisaacs) [inline interview-questions.md:64]: Totally, but let's think about this as if a interviewer asked you the question: "Give a brief description of proper pair programming techniques. What are the roles of each person?"

You answered what the description is, but you left them hanging on actually telling them what the role names are!

File: interview-questions.md [hunk 20]
```diff
@@ -8,62 +8,105 @@ Then, research the question to expand on your answer. Even if you feel you have
...
+
+ the first step is to describe what my function will do, what it is testing and what is the return.
+ the syntax is 
+ describe: function tested
+ "test: describe with your own words what's the scope of your code
+ expect: name of the function
+ toEqual: output of the function tested
+
+ Followed by the actual code with the function tested
+ running in the terminal my js file I will get a message (red/green) describing if test was passed or failed and give indications of the issue
```
Comment (andeeisaacs) [inline interview-questions.md:84]: Excellent! I think it would be wise to also include that "TDD" stands for "Test Driven Development"

File: testing-with-jest.test.js [hunk 21]
```diff
@@ -9,19 +9,73 @@
 // --------------------1a) Create a test for a function called addThemUp that takes two numbers as arguments and returns the sum.
 
 
+describe("addThemUpNo", () => {
+    test( "add two numbers" , () => {
+      expect(addThemUp()).toEqual(16)
```
Comment (andeeisaacs) [inline testing-with-jest.test.js:14]: Almost! We need our expect to take in two numbers. So we would give it test numbers ourself, such as 10, 5.
`expect(addThemUp(10, 5)).toEqual(16)`

File: testing-with-jest.test.js [hunk 22]
```diff
@@ -9,19 +9,73 @@
...
+        })
+})
+
+var numb1 = 10
+var numb2 = 5
+
+const addThemUpNo =  () => {
+    { return numb1 + numb2} 
+}
```
Comment (andeeisaacs) [inline testing-with-jest.test.js:23]: This, unfortunately, is not correct. 

First, the function is named incorrectly. Second does not have any arguments, we would need to provide the function arguments like so:
`const addThemUp = (num1, num2) => { logichere }`

Third, you wrapped your return in curly braces, which I'm not sure why. 
Fourth, your return hard codes the variables you created above, which are unnecessary and also makes the function ONLY work with those two variables. Instead, your return should look something like:
`return num1 + num2` as those would be the arguments being passed into it.

File: testing-with-jest.test.js [hunk 23]
```diff
@@ -9,19 +9,73 @@
...
+      expect(addThemUp()).toEqual(15)
+        })
+})
 
+var numb1 = 10
+var numb2 = 5
+
+const addThemUp =  () => {
+    { return numb1 + numb2} 
+}
```
Comment (andeeisaacs) [inline testing-with-jest.test.js:42]: This section, as the formatting implies, is where the actual code should have been for your function. You've just copy pasted the same test and code from above. 

Remember the steps to TDD.
First: Write the test
Second: See the test fail.
Third: Write the code.
Fourth: See the test pass.

When running these tests at the end, did the test pass?

File: testing-with-jest.test.js [hunk 24]
```diff
@@ -9,19 +9,73 @@
...
 
 
 
 // --------------------2a) Create a test for a function called triangleArea that takes the base and the height of a triange and returns the area. The area of a triange is the base times the height divided by two.
 
 
 
+describe("triAreaNo", () => {
+    test ("calculate a triangle area", () => {
+      expect(triAreaNo()).toEqual(50)
```
Comment (andeeisaacs) [inline testing-with-jest.test.js:52]: Again, our expect does not have any numbers in its argument.

File: testing-with-jest.test.js [hunk 25]
```diff
@@ -9,19 +9,73 @@
...
+})
+
+//it wil fail, the var babNo has the wrong value
+
+var babNo = 100
+var height = 10
+
+const triAreaNo = () => {
+    { return ((babNo * height)/2)} 
+}
```
Comment (andeeisaacs) [inline testing-with-jest.test.js:64]: This is still not passing arguments into its function. We also are not naming our function as the challenge is asking. We MUST be diligent in making sure our function names are exactly as our challenge is asking.

This is important because lets say we were working with existing code and you were tasked to create the tests. If you were renaming your functions to something else, you would break the entire existing code. Additionally, your tests wont run or pass, because your function names are different from your expect names.

**Review by andeeisaacs — APPROVED** [review]
Filippo, good attempt on your first week assessment! 

I want to encourage you to do the following:
- Take time to understand the concept of functions, the syntax and how they are reusable.
- Take time to understand the difference between a return and a console.log and WHY we do not use console.log's as our return
- .filter and .map, their syntax and how they operate.
- Take time to double check your challenge requirements. Many times you missed that the code needed to be wrapped in a function, or that the function name needed to be named a specific way. 

Overall, while there is definitely room to improve, you did great for your first week. If you take the time to really understand the things above, and double, triple check your challenge requirements, you'll be killing it with your code in no time! 

Feel free to merge!
